Update RobinhoodHashTable to validate its key in release builds

cdumez opened the pull request:
#### 17169d32e21511adf937e578e812e76f574500c1
<pre>
Update RobinhoodHashTable to validate its key in release builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=288207">https://bugs.webkit.org/show_bug.cgi?id=288207</a>

Reviewed by Ryosuke Niwa.

Update RobinhoodHashTable to validate its key in release builds, as security hardening.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::CachedHashMap::encode):
(JSC::CachedHashMap::decode const):
* Source/WTF/wtf/RobinHoodHashMap.h:
* Source/WTF/wtf/RobinHoodHashTable.h:
(WTF::shouldValidateKey&gt;::checkKey):
(WTF::shouldValidateKey&gt;::lookup):
(WTF::shouldValidateKey&gt;::inlineLookup):
(WTF::shouldValidateKey&gt;::initializeBucket):
(WTF::shouldValidateKey&gt;::add):
(WTF::shouldValidateKey&gt;::maintainProbeDistanceForAdd):
(WTF::shouldValidateKey&gt;::addPassingHashCode):
(WTF::shouldValidateKey&gt;::reinsert):
(WTF::shouldValidateKey&gt;::find):
(WTF::shouldValidateKey&gt;::find const):
(WTF::shouldValidateKey&gt;::contains const):
(WTF::shouldValidateKey&gt;::removeAndInvalidateWithoutEntryConsistencyCheck):
(WTF::shouldValidateKey&gt;::removeAndInvalidate):
(WTF::shouldValidateKey&gt;::remove):
(WTF::shouldValidateKey&gt;::removeWithoutEntryConsistencyCheck):
(WTF::shouldValidateKey&gt;::allocateTable):
(WTF::shouldValidateKey&gt;::deallocateTable):
(WTF::shouldValidateKey&gt;::expand):
(WTF::shouldValidateKey&gt;::computeBestTableSize):
(WTF::shouldValidateKey&gt;::shrinkToBestSize):
(WTF::shouldValidateKey&gt;::rehash):
(WTF::shouldValidateKey&gt;::clear):
(WTF::shouldValidateKey&gt;::RobinHoodHashTable):
(WTF::shouldValidateKey&gt;::swap):
(WTF::=):
(WTF::shouldValidateKey&gt;::checkTableConsistency const):
(WTF::shouldValidateKey&gt;::checkTableConsistencyExceptSize const):
(WTF::SizePolicy&gt;::checkKey): Deleted.
(WTF::SizePolicy&gt;::lookup): Deleted.
(WTF::SizePolicy&gt;::inlineLookup): Deleted.
(WTF::SizePolicy&gt;::initializeBucket): Deleted.
(WTF::SizePolicy&gt;::add): Deleted.
(WTF::SizePolicy&gt;::maintainProbeDistanceForAdd): Deleted.
(WTF::SizePolicy&gt;::addPassingHashCode): Deleted.
(WTF::SizePolicy&gt;::reinsert): Deleted.
(WTF::SizePolicy&gt;::find): Deleted.
(WTF::SizePolicy&gt;::find const): Deleted.
(WTF::SizePolicy&gt;::contains const): Deleted.
(WTF::SizePolicy&gt;::removeAndInvalidateWithoutEntryConsistencyCheck): Deleted.
(WTF::SizePolicy&gt;::removeAndInvalidate): Deleted.
(WTF::SizePolicy&gt;::remove): Deleted.
(WTF::SizePolicy&gt;::removeWithoutEntryConsistencyCheck): Deleted.
(WTF::SizePolicy&gt;::allocateTable): Deleted.
(WTF::SizePolicy&gt;::deallocateTable): Deleted.
(WTF::SizePolicy&gt;::expand): Deleted.
(WTF::SizePolicy&gt;::computeBestTableSize): Deleted.
(WTF::SizePolicy&gt;::shrinkToBestSize): Deleted.
(WTF::SizePolicy&gt;::rehash): Deleted.
(WTF::SizePolicy&gt;::clear): Deleted.
(WTF::SizePolicy&gt;::RobinHoodHashTable): Deleted.
(WTF::SizePolicy&gt;::swap): Deleted.
(WTF::SizePolicy&gt;::checkTableConsistency const): Deleted.
(WTF::SizePolicy&gt;::checkTableConsistencyExceptSize const): Deleted.

Canonical link: <a href="https://commits.webkit.org/290890@main">https://commits.webkit.org/290890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50ac51f99254477b6f267e012a7f0ef701c27a3f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91164 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/202 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96165 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41920 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11111 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70063 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27580 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82599 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50389 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8239 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/172 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41057 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83985 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78559 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/188 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98148 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89932 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18365 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13466 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79065 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78268 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19389 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22780 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/127 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11552 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18366 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23680 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112499 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18089 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32661 "Found 13 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-medium.js.mini-mode, microbenchmarks/memcpy-wasm-small.js.no-llint, microbenchmarks/memcpy-wasm.js.dfg-eager, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.default-wasm, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21550 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19865 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (commented):
r=me assuming perf A/B comes back negative.

cdumez committed:
Update RobinhoodHashTable to validate its key in release builds
https://bugs.webkit.org/show_bug.cgi?id=288207

Reviewed by Ryosuke Niwa.

Update RobinhoodHashTable to validate its key in release builds, as security hardening.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::CachedHashMap::encode):
(JSC::CachedHashMap::decode const):
* Source/WTF/wtf/RobinHoodHashMap.h:
* Source/WTF/wtf/RobinHoodHashTable.h:
(WTF::shouldValidateKey>::checkKey):
(WTF::shouldValidateKey>::lookup):
(WTF::shouldValidateKey>::inlineLookup):
(WTF::shouldValidateKey>::initializeBucket):
(WTF::shouldValidateKey>::add):
(WTF::shouldValidateKey>::maintainProbeDistanceForAdd):
(WTF::shouldValidateKey>::addPassingHashCode):
(WTF::shouldValidateKey>::reinsert):
(WTF::shouldValidateKey>::find):
(WTF::shouldValidateKey>::find const):
(WTF::shouldValidateKey>::contains const):
(WTF::shouldValidateKey>::removeAndInvalidateWithoutEntryConsistencyCheck):
(WTF::shouldValidateKey>::removeAndInvalidate):
(WTF::shouldValidateKey>::remove):
(WTF::shouldValidateKey>::removeWithoutEntryConsistencyCheck):
(WTF::shouldValidateKey>::allocateTable):
(WTF::shouldValidateKey>::deallocateTable):
(WTF::shouldValidateKey>::expand):
(WTF::shouldValidateKey>::computeBestTableSize):
(WTF::shouldValidateKey>::shrinkToBestSize):
(WTF::shouldValidateKey>::rehash):
(WTF::shouldValidateKey>::clear):
(WTF::shouldValidateKey>::RobinHoodHashTable):
(WTF::shouldValidateKey>::swap):
(WTF::=):
(WTF::shouldValidateKey>::checkTableConsistency const):
(WTF::shouldValidateKey>::checkTableConsistencyExceptSize const):
(WTF::SizePolicy>::checkKey): Deleted.
(WTF::SizePolicy>::lookup): Deleted.
(WTF::SizePolicy>::inlineLookup): Deleted.
(WTF::SizePolicy>::initializeBucket): Deleted.
(WTF::SizePolicy>::add): Deleted.
(WTF::SizePolicy>::maintainProbeDistanceForAdd): Deleted.
(WTF::SizePolicy>::addPassingHashCode): Deleted.
(WTF::SizePolicy>::reinsert): Deleted.
(WTF::SizePolicy>::find): Deleted.
(WTF::SizePolicy>::find const): Deleted.
(WTF::SizePolicy>::contains const): Deleted.
(WTF::SizePolicy>::removeAndInvalidateWithoutEntryConsistencyCheck): Deleted.
(WTF::SizePolicy>::removeAndInvalidate): Deleted.
(WTF::SizePolicy>::remove): Deleted.
(WTF::SizePolicy>::removeWithoutEntryConsistencyCheck): Deleted.
(WTF::SizePolicy>::allocateTable): Deleted.
(WTF::SizePolicy>::deallocateTable): Deleted.
(WTF::SizePolicy>::expand): Deleted.
(WTF::SizePolicy>::computeBestTableSize): Deleted.
(WTF::SizePolicy>::shrinkToBestSize): Deleted.
(WTF::SizePolicy>::rehash): Deleted.
(WTF::SizePolicy>::clear): Deleted.
(WTF::SizePolicy>::RobinHoodHashTable): Deleted.
(WTF::SizePolicy>::swap): Deleted.
(WTF::SizePolicy>::checkTableConsistency const): Deleted.
(WTF::SizePolicy>::checkTableConsistencyExceptSize const): Deleted.

Canonical link: https://commits.webkit.org/290890@main
